pkg/fs/dir.go

```diff
@@ -85,31 +91,61 @@ type ObjectInfo struct {
 	Err          error
 }
 
-// readDir - read 'scanDir' directory.  It returns list of ObjectInfo where
-//           each object name is appended with 'namePrefix'
-func readDir(scanDir, namePrefix string) (objInfos []ObjectInfo) {
+// Using search.Sort() internally to jump to the file entry containing the prefix
+func searchFileInfos(fileInfos []os.FileInfo, x string) int {
+	processFunc := func(i int) bool {
+		return fileInfos[i].Name() >= x
+	}
+	return sort.Search(len(fileInfos), processFunc)
+}
+
+// ReadDir - read 'scanDir' directory.  It returns list of ObjectInfo.
+// Each object name is appended with 'namePrefix'.
+func readDir(scanDir, namePrefix, queryPrefix string, isFirst bool) (objInfos []ObjectInfo) {
 	f, err := os.Open(scanDir)
 	if err != nil {
 		objInfos = append(objInfos, ObjectInfo{Err: err})
 		return
 	}
-
 	fis, err := f.Readdir(-1)
 	if err != nil {
 		f.Close()
 		objInfos = append(objInfos, ObjectInfo{Err: err})
 		return
 	}
-
-	// Close the directory
+	// Close the directory.
 	f.Close()
-
 	// Sort files by Name.
 	sort.Sort(byName(fis))
 
-	// Populate []ObjectInfo from []FileInfo
+	var prefixIndex int
+	// Searching for entries with objectName containing  prefix.
+	// Binary search is used  for efficient search.
+	if queryPrefix != "" && isFirst {
+		prefixIndex = searchFileInfos(fis, queryPrefix)
+		if prefixIndex == len(fis) {
+			return
+		}
+
+		if !strings.HasPrefix(fis[prefixIndex].Name(), queryPrefix) {
+			return
+		}
+		fis = fis[prefixIndex:]
+
+	}
+
+	// Populate []ObjectInfo from []FileInfo.
 	for _, fi := range fis {
 		name := fi.Name()
+		if queryPrefix != "" && isFirst {
+			// If control is here then there is a queryPrefix, and there are objects which satisfies the prefix.
+			// Since the result is sorted, the object names which satisfies query prefix would be stored one after the other.
+			// Push the objectInfo only if its contains the prefix.
+			// This ensures that the channel containing object Info would only has objects with the given queryPrefix.
+			if !strings.HasPrefix(name, queryPrefix) {
```

Comment:
Is this a double verification since a similar check is done before populating 'fis' ?